### PR TITLE
Add more questions to basic Next.js tutorial.

### PR DESCRIPTION
**Pull request by leerob**
I ran through the entire basic tutorial with a fresh set of eyes. Here's what changed:

- Adds 15 new questions to the basic tutorial
- Updates links mentioning edge network to point to documentation
- Small rewording of phrases for clarity
- Added a few more links when introducing new acronyms and methods
- Added a few notes to help clarify new concepts

### CODE REVIEW

File: pages/learn/basics/api-routes/index.mdx [hunk 1]
```diff
@@ -6,7 +6,8 @@ export const meta = {
   lessonId: 'api-routes'
 }
 
-Next.js has support for [API Routes](/docs/api-routes/introduction), which lets you easily create an API endpoint as a Node.js function. Although it’s not necessary for our blog app, we’ll briefly talk about how to use it in this lesson.
+Next.js has support for [API Routes](/docs/api-routes/introduction), which let you easily create an API endpoint as a Node.js serverless function.
+Although it’s not necessary for our blog app, we’ll briefly talk about how to use it in this lesson.
```
Comment (lfades) [inline pages/learn/basics/api-routes/index.mdx:10]: Nit: Markdown ignores these new lines, it will continue to be a single line so it's better to have it as one line in the code to avoid confusion.

File: pages/learn/basics/data-fetching/getstaticprops-details.mdx [hunk 2]
```diff
@@ -15,8 +19,6 @@ Here is some essential information you should know about [`getStaticProps`](/doc
 In `lib/posts.js`, we’ve implemented `getSortedPostsData` which fetches data from the file system. But you can fetch the data from other sources, like an external API endpoint, and it’ll work just fine:
 
 ```js
-import fetch from 'node-fetch'
-
```
Comment (lfades) [inline pages/learn/basics/data-fetching/getstaticprops-details.mdx:19]: Does this change requires a PR to next-learn-starter?

File: pages/learn/basics/dynamic-routes/render-markdown.mdx [hunk 3]
```diff
@@ -59,7 +59,9 @@ export async function getStaticProps({ params }) {
 }
 ```
 
-Finally, update the `Post` component to render `contentHtml` using `dangerouslySetInnerHTML`:
+> **Note**: `async`/`await` allow you to fetch data [asynchronously](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function).
```
Comment (lfades) [inline pages/learn/basics/dynamic-routes/render-markdown.mdx:62]: I think this note should be on the note that says `Important` that's above

**Review by lfades — COMMENTED** [review]
I'm not a fan of changing the amount of answers. We could go for 4 answers per test, that way it will feel more like a test than a question (you don't have to answer questions, but a test is there to see if you learned, and I like that more)

File: pages/learn/basics/api-routes/index.mdx (hunk 1 above)
Comment (leerob) [reply pages/learn/basics/api-routes/index.mdx:10]: This might be personal preference, but it's easier to read the Markdown file when the lines aren't extremely long.

File: pages/learn/basics/data-fetching/getstaticprops-details.mdx (hunk 2 above)
Comment (leerob) [reply pages/learn/basics/data-fetching/getstaticprops-details.mdx:19]: Nope! https://github.com/vercel/next-learn-starter/search?q=node-fetch&type=code

**Comment by leerob** [comment]
> I'm not a fan of changing the amount of answers. We could go for 4 answers per test, that way it will feel more like a test than a question (you don't have to answer questions, but a test is there to see if you learned, and I like that more)

There are certain questions where it's not feasible to have four answers. I would prefer two relevant answers (that make it slightly more difficult to determine which is the right answer) then four answers where two are completely bogus. 

Thoughts?

File: pages/learn/basics/data-fetching/getstaticprops-details.mdx [hunk 4]
```diff
@@ -25,6 +27,8 @@ export async function getSortedPostsData() {
 }
 ```
 
+> **Note**: Next.js polyfills [`fetch()`](/docs/basic-features/supported-browsers-features#server-side-polyfills) in the Node.js environment. You don't need to import it.
```
Comment (chibicode) [inline pages/learn/basics/data-fetching/getstaticprops-details.mdx:30]: Next.js actually polyfills `fetch` in the client side as well: https://nextjs.org/blog/next-9-1-7#new-built-in-polyfills-fetch-url-and-objectassign

File: pages/learn/basics/assets-metadata-css/metadata.mdx [hunk 5]
```diff
@@ -6,8 +6,7 @@ export const meta = {
   title: 'Assets, Metadata, and CSS',
   courseId: 'basics',
   lessonId: 'assets-metadata-css',
-  stepId: 'metadata',
-  subtitle: 'Metadata'
```
Comment (chibicode) [inline pages/learn/basics/assets-metadata-css/metadata.mdx:10]: Why was this line removed?

File: pages/learn/basics/api-routes/index.mdx (hunk 1 above)
Comment (chibicode) [reply pages/learn/basics/api-routes/index.mdx:10]: I think there are a couple of points to think about:

- It’s harder to see what the diff is if you both change the text and split the lines.
- If you split the lines only for the changes you made, it introduces inconsistencies with the rest of the code. We’ll then end up having two different code styles. Ideally, if we’re going to split the lines, we should use [this prettier option](https://prettier.io/docs/en/options.html#prose-wrap) to enforce it.

File: pages/learn/basics/assets-metadata-css/layout-component.mdx [hunk 6]
```diff
@@ -6,7 +6,11 @@ export const meta = {
   courseId: 'basics',
   lessonId: 'assets-metadata-css',
   stepId: 'layout-component',
-  subtitle: 'Layout Component'
+  subtitle: 'Layout Component',
+  question: {
+    answers: ['CSS Modules', 'Sass', 'styled-jsx', 'styled-components'],
```
Comment (chibicode) [inline pages/learn/basics/assets-metadata-css/layout-component.mdx:11]: This is fine for now. But eventually, I’d like to get rid of styled-jsx altogether from the lessons and from the starter template: https://github.com/vercel/next-learn-starter/blob/master/learn-starter/pages/index.js

The starter template is a simplified version of the `create-next-app` template (basically I removed `pages/api` from it: https://github.com/vercel/next.js/tree/canary/packages/create-next-app/templates/default). When I created this tutorial, the `create-next-app` template was also using `styled-jsx`, but since then styled-jsx was removed and it only uses CSS modules. I think we should do so for the tutorial as well and completely remove mentions of styled-jsx. But that can be a separate PR.

File: pages/learn/basics/data-fetching/request-time.mdx [hunk 7]
```diff
@@ -6,7 +6,15 @@ export const meta = {
...
   stepId: 'request-time',
-  subtitle: 'Fetching Data at Request Time'
+  subtitle: 'Fetching Data at Request Time',
+  question: {
+    answers: [
+      `When you can pre-render the data ahead of a user's request`,
+      'When you need to fetch data at request time instead of build time',
+      'Private, user-specific pages where SEO is not relevant'
+    ],
+    correctAnswer: 'Private, user-specific pages where SEO is not relevant'
```
Comment (chibicode) [inline pages/learn/basics/data-fetching/request-time.mdx:16]: Actually, I think "When you need to fetch data at request time instead of build time" is also a correct answer as well

File: pages/learn/basics/data-fetching/two-forms.mdx [hunk 8]
```diff
@@ -6,7 +6,15 @@ export const meta = {
...
   stepId: 'two-forms',
-  subtitle: 'Two Forms of Pre-rendering'
+  subtitle: 'Two Forms of Pre-rendering',
+  question: {
+    answers: [
+      'When you have thousands of pages',
+      'When the data needs to be up-to-date with every request',
+      'When you are building a page specific to a user'
+    ],
+    correctAnswer: 'When the data needs to be up-to-date with every request'
```
Comment (chibicode) [inline pages/learn/basics/data-fetching/two-forms.mdx:16]: I think SSR can also be used to build a page specific to a user, so this can be correct as well

File: pages/learn/basics/dynamic-routes/implement-getstaticprops.mdx [hunk 9]
```diff
@@ -111,6 +118,8 @@ Again, here’s the graphical summary of what we’ve done:
   height={1054 / 2}
 />
 
-But we still haven’t displayed the blog **markdown content**. So let’s do this next.
+We still haven’t displayed the blog **markdown content**. Let’s do this next.
+
+**Question**: How did `getStaticProps` know the variable was named `id`?
```
Comment (chibicode) [inline pages/learn/basics/dynamic-routes/implement-getstaticprops.mdx:123]: It’s hard to remember what "the variable" is referring to - the code snippet for `getStaticProps` appears much earlier in the page, so people would have to scroll the page up quite a bit to read the code again.

File: pages/learn/basics/dynamic-routes/page-path-external-data.mdx [hunk 10]
```diff
@@ -45,7 +45,8 @@ export default function Post() {
 }
 ```
 
-Now, here’s what’s new: We’ll export an async function called [`getStaticPaths`](/docs/basic-features/data-fetching#getstaticpaths-static-generation) from this page. In this function, we need to return a list of **possible values** for `id`.
+Now, here’s what’s new: We’ll export an async function called [`getStaticPaths`](/docs/basic-features/data-fetching#getstaticpaths-static-generation) from this page.
+In this function, we need to return a list of **possible values** for `id`. .
```
Comment (chibicode) [inline pages/learn/basics/dynamic-routes/page-path-external-data.mdx:49]:
```suggestion
In this function, we need to return a list of **possible values** for `id`.
```

File: pages/learn/basics/dynamic-routes/page-path-external-data.mdx [hunk 11]
```diff
@@ -59,7 +60,8 @@ export async function getStaticPaths() {
 }
 ```
 
-Finally, we need to implement [`getStaticProps`](/docs/basic-features/data-fetching#getstaticprops-static-generation) again - this time, to fetch necessary data for the blog post with a given `id`. [`getStaticProps`](/docs/basic-features/data-fetching#getstaticprops-static-generation) is given `params`, which contains `id`.
+Finally, we need to implement [`getStaticProps`](/docs/basic-features/data-fetching#getstaticprops-static-generation) again - this time, to fetch necessary data for the blog post with a given `id`.
+[`getStaticProps`](/docs/basic-features/data-fetching#getstaticprops-static-generation) is given `params`, which contains `id` (based on variable defined in the file name).
```
Comment (chibicode) [inline pages/learn/basics/dynamic-routes/page-path-external-data.mdx:64]: What about:

```suggestion
[`getStaticProps`](/docs/basic-features/data-fetching#getstaticprops-static-generation) is given `params`, which contains `id` (because the file name is `[id].js`).
```

File: pages/learn/basics/navigate-between-pages/client-side.mdx [hunk 12]
```diff
@@ -54,4 +61,6 @@ You can learn more about the `Link` component [in the API reference for `next/li
 >
 > If you need to add attributes like, for example, `className`, add it to the `a` tag, _not_ to the `Link` tag. [Here’s an example](https://github.com/vercel/next-learn-starter/blob/master/snippets/link-classname-example.js).
 
+**Question**: When a user visits `/posts/first-post`, what JavaScript is loaded?
```
Comment (chibicode) [inline pages/learn/basics/navigate-between-pages/client-side.mdx:64]: This could lead to confusion if the reader misinterprets the question as:

"The user loaded the home page, and then clicked the link for `/posts/first-post` - what JavaScript is loaded?"

Because this application only has 2 pages (home page and `/posts/first-post`), if the user visits `/posts/first-post` from the home page, then they have loaded JS for both the home page and the `/posts/first-post` page, so the correct answer would be "JavaScript for the entire application is loaded."

"Only the JavaScript for /posts/first-post is loaded." will be the correct answer only if the user loads `/posts/first-post` as the first loaded page.

File: pages/learn/basics/navigate-between-pages/pages-in-nextjs.mdx [hunk 13]
```diff
@@ -6,7 +6,11 @@ export const meta = {
   courseId: 'basics',
   lessonId: 'navigate-between-pages',
   stepId: 'pages-in-nextjs',
-  subtitle: 'Pages in Next.js'
+  subtitle: 'Pages in Next.js',
+  question: {
+    answers: ['/authors/me.js', 'pages/authors/me.js', 'routes/authors/me.js'],
```
Comment (chibicode) [inline pages/learn/basics/navigate-between-pages/pages-in-nextjs.mdx:11]:
```suggestion
    answers: ['authors/me.js', 'pages/authors/me.js', 'routes/authors/me.js'],
```

**Comment by chibicode** [comment]
@lfades 

> There are certain questions where it's not feasible to have four answers. I would prefer two relevant answers (that make it slightly more difficult to determine which is the right answer) then four answers where two are completely bogus.

I agree with @leerob. I don’t think we should have unnecessary noise in question choices. Adding completely bogus answer choices makes the tutorial sound condescending in my opinion.

**Comment by chibicode** [comment]
@leerob thank you for the changes! One request though: I’ve written the final paragraphs of each lesson such that it flows to the next page smoothly (in the following screenshots, I added the word "Let’s do this next")

⬇️ ⬇️ ⬇️ ⬇️ ⬇️ ⬇️

Screen Shot 2020-09-18 at 2 25 12 AM

⬆️ ⬆️ ⬆️ ⬆️ ⬆️ ⬆️ 
 
But I think adding another paragraph below it, starting the word "Question:" kind of weakens the previous paragraph too much.

## Suggestion

What do you think about visually separating the questions by wrapping the question with a box? (I used `#fafafa` background and `#eaeaea` border)  By doing this, the question itself won’t become the final paragraph:

⬇️ ⬇️ ⬇️ ⬇️ ⬇️ ⬇️

Screen Shot 2020-09-18 at 2 45 19 AM

⬆️ ⬆️ ⬆️ ⬆️ ⬆️ ⬆️ 

I also used the word "Quick Review" instead of "Question" - I think it sounds better, but let me know what you think.

File: pages/learn/basics/data-fetching/getstaticprops-details.mdx (hunk 4 above)
Comment (leerob) [reply pages/learn/basics/data-fetching/getstaticprops-details.mdx:30]: Is it confusing to mention that, though, when talking about the server? That was why I originally just mentioned in the Node context. Should I just say "Next.js polyfills fetch() on both the client and server. You don't need to import it"?

File: pages/learn/basics/assets-metadata-css/metadata.mdx (hunk 5 above)
Comment (leerob) [reply pages/learn/basics/assets-metadata-css/metadata.mdx:10]: Accident – good catch. Will re-add!

File: pages/learn/basics/data-fetching/two-forms.mdx (hunk 8 above)
Comment (leerob) [reply pages/learn/basics/data-fetching/two-forms.mdx:16]: Agreed. Writing wrong answers is hard 😓  Any suggestion on how to make this more clear?

File: pages/learn/basics/dynamic-routes/implement-getstaticprops.mdx (hunk 9 above)
Comment (leerob) [reply pages/learn/basics/dynamic-routes/implement-getstaticprops.mdx:123]: I will try to reword this! It's a common piece of feedback I've received that people struggle to make the connection between the filename and where this magic variable comes from, which is why I wanted to include this specifically.

File: pages/learn/basics/api-routes/index.mdx (hunk 1 above)
Comment (leerob) [reply pages/learn/basics/api-routes/index.mdx:10]: Those are good points. I will revert this and we can discuss in another PR is we want to add that Prettier option 👍

File: pages/learn/basics/data-fetching/getstaticprops-details.mdx (hunk 4 above)
Comment (leerob) [reply pages/learn/basics/data-fetching/getstaticprops-details.mdx]: Updated to reflect both 👍

File: pages/learn/basics/navigate-between-pages/client-side.mdx (hunk 12 above)
Comment (leerob) [reply pages/learn/basics/navigate-between-pages/client-side.mdx]: Reworded this: A user opens their browser and navigates to `your-blog.com/posts/first-post`. What JavaScript is initially loaded for this page?

**Comment by leerob** [comment]
@chibicode Re: adding an outline to the question box. I like the idea – is there any easy way to keep the Markdown transformation with bold / syntax highlighting? Right now, the question text is outside the `AnswerBox`, so MDX is still processing it. If we were to move the question text to `meta` and the forward it to the component, it wouldn't get transformed into JSX and instead be a raw string. 

I also thought about directly importing `AnswerBox` as a component inside the MDX file, but that might get more complicated with tracking the user's record. 

image

Thoughts?

**Comment by lfades** [comment]
@leerob You can try the following:

Add a custom element around the question, iirc if you add markdown inside that element MDX should parse it first before sending it to the custom component:

```jsx
<Question>
  Something about `page/index.js`
<Question />
```

If that one doesn't work this one will:

```jsx
<Question>

Something about `page/index.js`

<Question />
```

> In the second case `Question` will receive a paragraph, instead of multiple children with text

Then in the `Question` component you can render the answers (assign the meta to a context provider so the element can get it)

**Comment by leerob** [comment]
> I also thought about directly importing AnswerBox as a component inside the MDX file, but that might get more complicated with tracking the user's record.

Turns out hooks make this really easy. I was overthinking it 😄  Thanks for the info @lfades!

**Comment by lfades** [comment]
@leerob Also think about how will it work in the future, if we somehow end up needing the answer in the meta for the certificate, maybe having it outside may make it harder? but that's just my nitpicking btw

**Comment by leerob** [comment]
Would you mind re-reviewing @lfades @chibicode?

**Review by lfades — APPROVED** [review]
LGTM

**Review by chibicode — APPROVED** [review]
